pkg/api/types.go

```diff
@@ -2119,34 +2119,121 @@ const (
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
+}
+
+type ComponentPhase string
+
+// These are the valid phases of a component.
+const (
+	// ComponentPending means the component has been created, but is still starting up.
+	ComponentPending ComponentPhase = "Pending"
+	// ComponentRunning means the component has started, but may or may not be healthy.
+	ComponentRunning ComponentPhase = "Running"
+	// ComponentTerminated means the component has been removed from the cluster, whether cleanly or not.
+	ComponentTerminated ComponentPhase = "Terminated"
+)
+
 type ComponentConditionType string
 
-// These are the valid conditions for the component.
+// These are the valid conditions for a component to have.
 const (
-	ComponentHealthy ComponentConditionType = "Healthy"
+	// ComponentRunningHealthy means the component is up and running, when true
+	ComponentRunningHealthy ComponentConditionType = "Healthy"
+	// ComponentTerminatedCleanly means the component is down and stopped without error, when true
+	ComponentTerminatedCleanly ComponentConditionType = "Cleanly"
 )
 
 type ComponentCondition struct {
-	Type    ComponentConditionType `json:"type"`
-	Status  ConditionStatus        `json:"status"`
-	Message string                 `json:"message,omitempty"`
-	Error   string                 `json:"error,omitempty"`
+	// Type of condition: Pending, Running, or Terminated
+	Type ComponentConditionType `json:"type"`
+	// Status of the condition: True, False, or Unknown
+	Status ConditionStatus `json:"status"`
+	// Reason (brief) for the transition to this condition
+	Reason string `json:"message,omitempty"`
+	// Message (detailed) explaining the reason for the transition to this condition
+	Message string `json:"error,omitempty"`
 }
 
-// ComponentStatus (and ComponentStatusList) holds the cluster validation info.
+// ComponentStatus is information about the current status of a component.
 type ComponentStatus struct {
+	// ComponentPhase is the current lifecycle phase of the component.
+	Phase ComponentPhase `json:"phase"`
+	// Conditions is an array of historical node conditions.
+	Conditions []ComponentCondition `json:"conditions,omitempty"`
+	// LastUpdateTime is the last time the component status was updated.
+	// TODO(karlkfi): move LastUpdateTime to ObjectMeta? maybe both (one for spec, one for status)?
+	LastUpdateTime util.Time `json:"lastUpdateTime,omitempty"`
```

Comment:
I'm not sure I follow. ComponentCondition doesn't have any timestamps. Are you implying it should? Even if it did, you'd have to iterate through them to find the latest to interpolate LastUpdateTime. And you'd have to validate that the Conditions array can't be empty.

TBH, I'm not really sure why Conditions is a list in the other APIs. It seems like it should be a map with ConditionType as the unique key.

As is:
- LastUpdateTime is the last time we heard a heartbeat (status update) from the Component, the last time we are sure it was in this Phase/Condition. 
  - Required in order to timeout and transition to unheathly/lost/terminated.
- LastTransitionTime is the last time Phase or Condition changed.
  - Required in order to detect thrashing.